templates/template_processor.go

```diff
@@ -22,12 +24,23 @@ import (
 // The name of the variable that contains the current value of the loop in each iteration of for_each
 const eachVarName = "__each__"
 
+// Maintaining original function name and signature for backwards compatability
+func ProcessTemplate(options, rootOpts *options.BoilerplateOptions, thisDep variables.Dependency) error {
+	_, err := processTemplateInternal(options, rootOpts, thisDep, false)
+	return err
+}
+
+// ProcessTemplateWithManifest works like ProcessTemplate but also returns the manifest of generated files
+func ProcessTemplateWithManifest(options, rootOpts *options.BoilerplateOptions, thisDep variables.Dependency) (*manifest.Manifest, error) {
+	return processTemplateInternal(options, rootOpts, thisDep, true)
+}
+
 // Process the boilerplate template specified in the given options and use the existing variables. This function will
 // download remote templates to a temporary working directory, which is cleaned up at the end of the function. This
 // function will load any missing variables (either from command line options or by prompting the user), execute all the
 // dependent boilerplate templates, and then execute this template. Note that we pass in rootOptions so that template
 // dependencies can inspect properties of the root template.
-func ProcessTemplate(options, rootOpts *options.BoilerplateOptions, thisDep variables.Dependency) error {
+func processTemplateInternal(options, rootOpts *options.BoilerplateOptions, thisDep variables.Dependency, returnManifest bool) (*manifest.Manifest, error) {
```

Comment:
There's a more idiomatic way to do this, and I think it's worth breaking backwards compatibility to achieve it.

I'd recommend creating a `NewTemplateProcessor` function that returns a template processor, accepting different options to configure it, then having a `Process()` method to do what it needs to do.

This is an example of that:
https://github.com/gruntwork-io/terragrunt/blob/main/internal/discovery/discovery.go#L107

It also allows us to get out of passing around global options like that.

I'm not the lead maintainer of this project, so I'll solicit some feedback from other engineers at Gruntwork. I recommend waiting for their input before making any changes to this PR.

Reply:
We've discussed this internally, and I can bless the decision to engage in a breaking change to make this codepath more maintainable by changing the API.

We'll look to merge this in with another anticipated breaking change to require confirmation before running shells/hooks for users when in interactive mode.

Reply:
This has been updated. I just need some feedback as per below regarding the versioned manifest. Should be able to update early next week.